### ISSUE TITLE
Simple Van Der Pol ODE not giving sensible results

### ISSUE THREAD

**Issue by kyrxanthos** [dynamicslab/pysindy#231]
I have been trying to use the package to recover the Van Der Pol Equations, but I get very poor results with the default parameters of SINDy and I was wondering what kind of things I should change. 

```
import numpy as np
import pysindy as ps
from scipy.integrate import odeint
import matplotlib.pyplot as plt
from scipy.integrate import solve_ivp

# Initialize integrator keywords for solve_ivp to replicate the odeint defaults
integrator_keywords = {}
integrator_keywords['rtol'] = 1e-12
integrator_keywords['method'] = 'LSODA'
integrator_keywords['atol'] = 1e-12

# Generate measurement data
dt = .01

t_train = np.arange(0, 15, dt)
x0_train = [1, 2]
t_train_span = (t_train[0], t_train[-1])

# Van der Pol ODE
def van_der_pol(t, x, p=[1]):
    return [-x[1], - p[0] * (1 - x[0] ** 2) * x[1] + x[0]]

x_train = solve_ivp(van_der_pol, t_train_span, x0_train,
                    t_eval=t_train, **integrator_keywords).y.T

# Instantiate and fit the SINDy model
model = ps.SINDy(feature_names=['x1', 'x2'])
model.fit(x_train, t=dt)
model.print()
model.score(x_train, dt)
```
The obtained ODE is: 

`(x1)' = -1.000 x2
(x2)' = 1.505 x1 + -0.820 x2 + 0.616 x1^2 + -0.116 x1 x2`

with a score of 0.97. Clearly this is quite far off from the actual system of equations. Testing it on a test set I get:

```
t_test = np.arange(0, 15, dt)
x0_test = np.array([0, 1.5])
t_test_span = (t_test[0], t_test[-1])
x_test = solve_ivp(van_der_pol, t_test_span, x0_test,
                   t_eval=t_test, **integrator_keywords).y.T
# Compare SINDy-predicted derivatives with finite difference derivatives
print('Model score: %f' % model.score(x_test, t=dt))

# Predict derivatives using the learned model
x_dot_test_predicted = model.predict(x_test)

# Compute derivatives with a finite difference method, for comparison
x_dot_test_computed = model.differentiate(x_test, t=dt)

fig, axs = plt.subplots(x_test.shape[1], 1, sharex=True, figsize=(7, 9))
for i in range(x_test.shape[1]):
    axs[i].plot(t_test, x_dot_test_computed[:, i],
                'k', label='numerical derivative')
    axs[i].plot(t_test, x_dot_test_predicted[:, i],
                'r--', label='model prediction')
    axs[i].legend()
    axs[i].set(xlabel='t', ylabel='$\dot x_{}$'.format(i))
fig.show()
```
![output](https://user-images.githubusercontent.com/67141092/180601631-a7cc400f-a4bd-4acb-b16b-096cff0ae596.png)

Any ideas of how I could improve my fit?

**Comment by Jacob-Stevens-Haas** [comment]
There's always several ways to try first:
* Different differentiation method.  The default, `FiniteDifference`, isn't usually the best, it's just the simplest.  Try something like smoothed finite difference.  Or, since you know the governing equations, try calculating the exact derivatives as `x_dot` and call `model.fit(x_train, t=dt, x_dot=x_dot)`.  If the recovered equations improve, then differentiation method is the culprit.  Similarly, shrinking dt will always help your derivative methods be more accurate.
* Different optimizer or optimizer parameters.  The default, STSLQ, works well but there are others. 
* Extend the time that you're generating data from 15 sec to much longer.

**Comment by kyrxanthos** [comment]
I have tried the `SmoothedFiniteDifference` differentiation method and other optimizers. The score does get better (goes up to 0.99) but the equations are still quite far off from the true Van Der Pol ones: 

```
(x1)' = -0.999 x2
(x2)' = 1.374 x1 + -0.291 x2 + 0.267 x1^2 + -0.134 x1 x2
```
I also extended the time and decreased the time steps but this did not have a large impact to the solution. I was wondering if the problem might be that the solution is very dependent on the initial conditions and if there is a way to add multiple time series with different initial conditions to produce a more accurate model. Other than that, is there anything fundamental I should change?

**Comment by akaptano** [comment]
I think your idea of trying to add other initial conditions is a good one. The Van Der Pol transient is hard to capture correctly  But the most obvious thing to try (that it appears you haven't) is to use a SINDy polynomial library that is up to cubic terms, since Van Der Pol is cubic in nonlinearity. The default is only quadratic in the state variables!

**Comment by kyrxanthos** [comment]
Indeed, increasing the degree of the polynomial library to 3 gave the correct solution. Thanks for the help!

**Comment by akaptano** [comment]
Good to hear, and glad you're using the library. 

Best,
Alan
